Discovery: DID registration by clients

reinkrul opened the pull request:
Uses https://github.com/nuts-foundation/nuts-node/pull/2711

This PR adds functionality and an API for discovery clients to register DIDs on discovery services. Presentations are automatically refreshed when they are about to expire.

woutslakhorst commented on discovery/interface.go:
```diff
@@ -95,4 +100,16 @@ type Server interface {
+	// If initial registration failed, it will return ErrRegistrationFailed, but it will keep retrying.
+	// If the function is called again for the same service/DID combination, it will try to refresh the registration.
+	// It returns an error if the service or DID is invalid/unknown.
+	StartRegistration(ctx context.Context, serviceID string, subjectDID did.DID) error
```
From an external point of view: `Register` and `Deregister`, the fact that it renews is unknown for the caller

woutslakhorst commented on docs/_static/discovery/v1.yaml:
```diff
@@ -71,6 +71,87 @@ paths:
+      
+        error returns:
+        * 400 - incorrect input: invalid/unknown service or DID.
+      operationId: startRegisteringPresentation
```
`register` and `deregister`

woutslakhorst commented on e2e-tests/discovery/run-test.sh:
```diff
@@ -0,0 +1,69 @@
+echo "---------------------------------------"
+echo "Searching for care organization registration..."
+echo "---------------------------------------"
+echo "TODO: Requires clients updating discovery service and search API (https://github.com/nuts-foundation/nuts-node/pull/2672)"
```
would be best to rebase and implement this?

reinkrul replied:
Clients are still not updating their local copy, working on that.

woutslakhorst commented on storage/sql_migrations/003_discoveryservice_client_registration.sql:
```diff
@@ -0,0 +1,20 @@
+    -- It comes from the service definition.
+    service_id        varchar(200) not null,
+    -- did is the DID that should be registered on the Discovery Service.
+    did               varchar      not null,
```
default for MySQL is 100, better to explicitly add the length

woutslakhorst commented on discovery/api/v1/wrapper.go:
```diff
@@ -83,3 +88,34 @@ func (w *Wrapper) RegisterPresentation(_ context.Context, request RegisterPresen
 }
+
+func (w *Wrapper) StartRegisteringPresentation(ctx context.Context, request StartRegisteringPresentationRequestObject) (StartRegisteringPresentationResponseObject, error) {
+	subjectDID, err := did.ParseDID(request.Did)
```
should check for isOwned, otherwise logs are spammed by default?

woutslakhorst commented on discovery/client.go:
```diff
@@ -0,0 +1,194 @@
+// It automatically refreshes registrations when they are about to expire.
+type registrationManager interface {
+	register(ctx context.Context, serviceID string, subjectDID did.DID) error
+	unregister(ctx context.Context, serviceID string, subjectDID did.DID) error
```
I prefer `deregister` because `unregistered` can both mean the change from registered to unregistered or the state unregistered.

woutslakhorst commented on discovery/cmd/cmd.go:
```diff
@@ -33,5 +33,8 @@ func FlagSet() *pflag.FlagSet {
 	flagSet.StringSlice("discovery.server.definition_ids", defs.Server.DefinitionIDs,
 		"IDs of the Discovery Service Definitions for which to act as server. "+
 			"If an ID does not map to a loaded service definition, the node will fail to start.")
+	flagSet.Duration("discovery.client.registration_refresh_interval", defs.Client.RegistrationRefreshInterval,
```
Explain duration format in description.

reinkrul commented on discovery/config.go:
```diff
@@ -35,14 +38,18 @@ type ServerConfig struct {
 	return Config{
 		Server: ServerConfig{},
+		Client: ClientConfig{
+			RegistrationRefreshInterval: 10 * time.Minute,
```
Arbitrary interval which can be configured

reinkrul commented on discovery/interface.go:
```diff
@@ -83,6 +85,9 @@ var ErrServiceNotFound = errors.New("discovery service not found")
 // but a presentation with this ID already exists.
 var ErrPresentationAlreadyExists = errors.New("presentation already exists")
 
+// ErrRegistrationFailed indicates registration of a presentation on a remote Discovery Service failed.
+var ErrRegistrationFailed = errors.New("registration failed")
```
Yes. Made the description a bit clearer.

reinkrul commented on storage/sql_migrations/003_discoveryservice_client_registration.sql:
```diff
@@ -0,0 +1,20 @@
+(
+    -- service_id is the ID of the Discover Service that the DID should be registered on.
+    -- It comes from the service definition.
+    service_id        varchar(200) not null,
```
I don't think so, we can always make it larger, but we need some limit i.m.o.
I think 200 characters is already plenty long.

reinkrul commented on discovery/client.go:
```diff
@@ -0,0 +1,194 @@
+	return instance
+}
+
+func (r *scheduledRegistrationManager) register(ctx context.Context, serviceID string, subjectDID did.DID) error {
```
Shall I rename client-side calls (registering a DID for a service, so VPs will be registered on the Discovery Service) to `activate(DID)`?

reinkrul replied:
Renamed stuff

gerardsn replied:
I agree with the renaming. It does not address this issue though.

reinkrul replied:
I swapped the 2 updates around; now it will store `0` first, making the presentation eligible for refreshing at the next interval. That way, if registration fails, it will always be retried as soon as possible (earlier, a db error could break the second call, delaying the retry). Downside is that it causes a functional race condition with `refresh()`, which could lead to 2 presentations being registered for the same DID short after each other. But this won't break anything, so this is the lesser evil.

reinkrul commented:
Changes to make things more understandable:
- Renamed client-side `registerDID` and `deregisterDID` to `activateServiceForDID` and `deactivateServiceForDID`
- Renamed `registrationManager` to `clientRegistrationManager`, `scheduledRegistrationManager` to `defaultClientRegistrationManager`
- Renamed `registrationManager`'s function to mirror `activate` rename
- Renamed `http.go` to `client.go`
- Renamed `wrapper.go` to `api.go`

reinkrul commented on discovery/module.go:
```diff
@@ -101,6 +116,7 @@ func (m *Module) Configure(_ core.ServerConfig) error {
 		}
 		m.serverDefinitions = serverDefinitions
 	}
+	m.httpClient = client.New(serverConfig.Strictmode, 10*time.Second, nil)
```
Maybe, maybe not. Time-out could be client-specific?
